File: pkg/geometry/triangle.go
package geometry

import "math"

// Triangle represents a triangular facet in 3D space
type Triangle struct {
	Normal Vector3
	V1, V2, V3 Vector3
}

// NewTriangle creates a new triangle
func NewTriangle(normal, v1, v2, v3 Vector3) Triangle {
	return Triangle{
		Normal: normal,
		V1:     v1,
		V2:     v2,
		V3:     v3,
	}
}

// CalculateNormal computes the normal vector for the triangle
func (t Triangle) CalculateNormal() Vector3 {
	edge1 := t.V2.Sub(t.V1)
	edge2 := t.V3.Sub(t.V1)
	return edge1.Cross(edge2).Normalize()
}

// Area returns the surface area of the triangle
func (t Triangle) Area() float64 {
	edge1 := t.V2.Sub(t.V1)
	edge2 := t.V3.Sub(t.V1)
	cross := edge1.Cross(edge2)
	return cross.Length() / 2.0
}

// EdgeLengths returns the lengths of all three edges
func (t Triangle) EdgeLengths() [3]float64 {
	return [3]float64{
		t.V1.Distance(t.V2),
		t.V2.Distance(t.V3),
		t.V3.Distance(t.V1),
	}
}

// Perimeter returns the total length of all edges
func (t Triangle) Perimeter() float64 {
	lengths := t.EdgeLengths()
	return lengths[0] + lengths[1] + lengths[2]
}

// Center returns the centroid of the triangle
func (t Triangle) Center() Vector3 {
	return Vector3{
		X: (t.V1.X + t.V2.X + t.V3.X) / 3.0,
		Y: (t.V1.Y + t.V2.Y + t.V3.Y) / 3.0,
		Z: (t.V1.Z + t.V2.Z + t.V3.Z) / 3.0,
	}
}

// Angles returns the three interior angles in radians
func (t Triangle) Angles() [3]float64 {
	e1 := t.V2.Sub(t.V1)
	e2 := t.V3.Sub(t.V2)
	e3 := t.V1.Sub(t.V3)

	// Angle at V1
	a1 := math.Acos(e1.Normalize().Dot(e3.Mul(-1).Normalize()))
	// Angle at V2
	a2 := math.Acos(e1.Mul(-1).Normalize().Dot(e2.Normalize()))
	// Angle at V3
	a3 := math.Acos(e2.Mul(-1).Normalize().Dot(e3.Normalize()))

	return [3]float64{a1, a2, a3}
}

File: GoSTL-Swift/GoSTL/Geometry/Triangle.swift
import Foundation
import simd

/// A triangle defined by three vertices
struct Triangle {
    var v1: Vector3
    var v2: Vector3
    var v3: Vector3
    var normal: Vector3

    // MARK: - Initializers

    init(v1: Vector3, v2: Vector3, v3: Vector3, normal: Vector3? = nil) {
        self.v1 = v1
        self.v2 = v2
        self.v3 = v3
        self.normal = normal ?? Self.calculateNormal(v1: v1, v2: v2, v3: v3)
    }

    // MARK: - Computed Properties

    /// Calculate the normal vector from vertices (cross product)
    static func calculateNormal(v1: Vector3, v2: Vector3, v3: Vector3) -> Vector3 {
        let edge1 = v2 - v1
        let edge2 = v3 - v1
        return edge1.cross(edge2).normalized()
    }

    /// Recalculate normal from current vertices
    mutating func updateNormal() {
        self.normal = Self.calculateNormal(v1: v1, v2: v2, v3: v3)
    }

    /// Area of the triangle
    func area() -> Double {
        let edge1 = v2 - v1
        let edge2 = v3 - v1
        return edge1.cross(edge2).length * 0.5
    }

    /// Length of each edge
    func edgeLengths() -> (Double, Double, Double) {
        let e1 = v1.distance(to: v2)
        let e2 = v2.distance(to: v3)
        let e3 = v3.distance(to: v1)
        return (e1, e2, e3)
    }

    /// Perimeter (sum of edge lengths)
    func perimeter() -> Double {
        let edges = edgeLengths()
        return edges.0 + edges.1 + edges.2
    }

    /// Centroid (center point)
    func center() -> Vector3 {
        (v1 + v2 + v3) / 3.0
    }

    /// Interior angles in radians
    func angles() -> (Double, Double, Double) {
        let a = v2 - v1
        let b = v3 - v1
        let c = v3 - v2

        let angle1 = acos(a.dot(b) / (a.length * b.length))
        let angle2 = acos((-a).dot(c) / (a.length * c.length))
        let angle3 = .pi - angle1 - angle2

        return (angle1, angle2, angle3)
    }

    /// All three vertices as an array
    var vertices: [Vector3] {
        [v1, v2, v3]
    }

    // MARK: - Ray Intersection

    /// Ray-triangle intersection using Möller–Trumbore algorithm
    /// - Parameter ray: The ray to test
    /// - Returns: Distance along the ray where intersection occurs, or nil if no intersection
    func intersect(ray: Ray) -> Float? {
        let epsilon: Float = 1e-6

        // Convert to Float for ray calculations
        let v1f = v1.float3
        let v2f = v2.float3
        let v3f = v3.float3

        let edge1 = v2f - v1f
        let edge2 = v3f - v1f
        let h = simd_cross(ray.direction, edge2)
        let a = simd_dot(edge1, h)

        // Ray is parallel to triangle
        if abs(a) < epsilon {
            return nil
        }

        let f = 1.0 / a
        let s = ray.origin - v1f
        let u = f * simd_dot(s, h)

        // Intersection point is outside triangle
        if u < 0.0 || u > 1.0 {
            return nil
        }

        let q = simd_cross(s, edge1)
        let v = f * simd_dot(ray.direction, q)

        // Intersection point is outside triangle
        if v < 0.0 || u + v > 1.0 {
            return nil
        }

        // Calculate t (distance along ray)
        let t = f * simd_dot(edge2, q)

        // Intersection is behind the ray origin
        if t < epsilon {
            return nil
        }

        return t
    }

    /// Get the point and normal at an intersection
    /// - Parameter ray: The ray to test
    /// - Returns: Tuple of (position, normal) if intersection occurs
    func intersectionPoint(ray: Ray) -> (position: Vector3, normal: Vector3)? {
        guard let t = intersect(ray: ray) else {
            return nil
        }
        let positionFloat = ray.origin + ray.direction * t
        let hitPoint = Vector3(Double(positionFloat.x), Double(positionFloat.y), Double(positionFloat.z))

        // Snap to nearest vertex
        let vertices = [v1, v2, v3]
        let distances = vertices.map { $0.distance(to: hitPoint) }
        let minIndex = distances.enumerated().min(by: { $0.element < $1.element })!.offset
        let snappedPosition = vertices[minIndex]

        return (snappedPosition, normal)
    }
}

// MARK: - Equatable

extension Triangle: Equatable {
    static func == (lhs: Triangle, rhs: Triangle) -> Bool {
        lhs.v1 == rhs.v1 &&
        lhs.v2 == rhs.v2 &&
        lhs.v3 == rhs.v3
    }
}

// MARK: - Codable

extension Triangle: Codable {
    enum CodingKeys: String, CodingKey {
        case v1, v2, v3, normal
    }
}

// MARK: - CustomStringConvertible

extension Triangle: CustomStringConvertible {
    var description: String {
        "Triangle(\(v1), \(v2), \(v3))"
    }
}
